### PR TITLE
Initial device-discovery

### PR DESCRIPTION
**Pull request by Mitko-Kerezov**
Positives
+ A new method in Yok for exposing a whole class publicly
+ Separation of device-discovery into android, ios and general device discovery
+ Extraction of a common-bootstrapper, used only by modules which require the common-lib
+ Android device discovery refactoring as an event-driven approach
+ Initial spike for iOS device detection

Negatives
+ our usage of the ref and ffi native dependencies has trouble with iojs 3.x

Follow-up
+ This pull request manages the code base only, a subsequent pull request will update the native dependencies

Ping @Fatme and @teobugslayer

### CODE REVIEW

**Comment by justcodebuilduser** [comment]
:heart:

File: bootstrap.ts [hunk 1]
```diff
@@ -1,8 +1,8 @@
 global._ = require("lodash");
 global.$injector = require("./yok").injector;
+$injector.require("options", "./options");
```
Comment (rosen-vladimirov) [inline bootstrap.ts]: this should not be here, it will not work for appbuilder-cli and nativescript-cli if you keep it this way

File: mobile/mobile-core/android-device-discovery.ts [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+	}
+
+	private deleteAndRemoveDevice(deviceIdentifier: string): void {
+		_.remove(this._devices, d => d === deviceIdentifier);
+		this.removeDevice(deviceIdentifier);
+	}
+
+	public startLookingForDevices(): IFuture<void> {
+		return(()=> {
+			let requestAllDevicesCommand = `"${this.$staticConfig.getAdbFilePath().wait()}" devices`;
```
Comment (rosen-vladimirov) [inline mobile/mobile-core/android-device-discovery.ts]: can we extract `this.$staticConfig.getAdbFilePath().wait()` to a variable and cache the result - we are using it here and in `ensureAdbServerStarted` as well

File: mobile/mobile-core/devices.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+	// TODO: add iosDeviceDiscovery as a dependency too
+	constructor(private $androidDeviceDiscovery:Mobile.IAndroidDeviceDiscovery) {
+		super();
+	}
+	
+	public initialize(): IFuture<void> {
+		return (() => {
+			this.$androidDeviceDiscovery.ensureAdbServerStarted().wait();
+			
+			setInterval(() => this.$androidDeviceDiscovery.startLookingForDevices(), 1000); // Maybe we should unref this as it will block CLI's execution and will hold the console.
```
Comment (rosen-vladimirov) [inline mobile/mobile-core/devices.ts]: have you checked this comment?

File: mobile/mobile-core/ios-device-discovery.ts [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+	}
+
+	public startLookingForDevices(): IFuture<void> {
+		return (() => {
+			this.subscribeForNotifications();
+			let defaultTimeoutInSeconds = 1;
+			let parsedTimeout =  this.$utils.getParsedTimeout(1);
+			let timeout = parsedTimeout > defaultTimeoutInSeconds ? parsedTimeout/1000 : defaultTimeoutInSeconds;
+			this.startRunLoopWithTimer(timeout);
+			setInterval(() => this.startRunLoopWithTimer(timeout), 1000); // I think we should unref this as it will block CLI's execution and will hold the console.
```
Comment (rosen-vladimirov) [inline mobile/mobile-core/ios-device-discovery.ts]: shouldn't we unref the runLoopTimer as described in the comment

File: mobile/mobile-core/ios-device-discovery.ts [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+			this.$logger.warn(this.error);
+		} else if(this.$hostInfo.isLinux) {
+			this.$logger.warn("In this version of the %s command-line interface, you cannot use connected iOS devices.", this.$staticConfig.CLIENT_NAME.toLowerCase());
+		}
+		
+		return Future.fromResult();
+	}
+}
+
+$injector.register("iOSDeviceDiscovery", ($errors: IErrors, $logger: ILogger, $fs: IFileSystem, $injector: IInjector, $iTunesValidator: Mobile.IiTunesValidator, $staticConfig: Config.IStaticConfig, $hostInfo: IHostInfo) => {
```
Comment (rosen-vladimirov) [inline mobile/mobile-core/ios-device-discovery.ts]: this `register` code must be refactored. Use the "classic" register way:
```TypeScript
$injector.register("iOSDeviceDiscovery", IOSDeviceDiscovery);
```
and move the logic from here directly in the IOSDeviceDiscovery class. Current code will not work very well with Proton setup

File: yok.ts [hunk 6]
```diff
@@ -59,7 +59,7 @@ import * as path from "path";
 let indent = "";
 function trace(formatStr: string, ...args: any[]) {
 	// uncomment following line when debugging dependency injection
-	//console.log(util.format(indent + formatStr, ...args));
+	// console.log(require('util').format(indent + formatStr, ...args));
```
Comment (rosen-vladimirov) [inline yok.ts]: require('util') ?
:crying_cat_face:

File: mobile/mobile-core/android-device-discovery.ts (hunk 2 above)
Comment (Mitko-Kerezov) [inline mobile/mobile-core/android-device-discovery.ts]: If you take a look at `this.$staticConfig.getAdbFilePath()`'s implementation, you can see it caches the value itself

Comment (rosen-vladimirov) [inline mobile/mobile-core/android-device-discovery.ts]: ok, this is not a merge stopper, I just don't like so long call in double quotes on two places in the same class:
```TypeScript
"${this.$staticConfig.getAdbFilePath().wait()}"
```

File: options.ts [hunk 7]
```diff
@@ -180,3 +180,4 @@ export class OptionsBase {
 	}
 
 }
+$injector.register("options", {});
```
Comment (rosen-vladimirov) [inline options.ts]: :crying_cat_face: 
please remove this line, this will break everything

**Comment by justcodebuilduser** [comment]
:broken_heart:

**Comment by justcodebuilduser** [comment]
:heart:

File: mobile/mobile-core/deviceEmitter.ts [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+			this.$androidDeviceDiscovery.on("deviceFound", (data: Mobile.IDevice) => {
+				this.emit("deviceFound", data.deviceInfo);
+			});
+
+			this.$androidDeviceDiscovery.on("deviceLost", (data: Mobile.IDevice) => {
+				this.emit("deviceLost", data.deviceInfo);
+			});
+
+			setInterval(() =>
+				fiberBootstrap.run(() => {
```
Comment (Fatme) [inline mobile/mobile-core/deviceEmitter.ts]: Why do you need to run this in fiber?

Comment (rosen-vladimirov) [inline mobile/mobile-core/deviceEmitter.ts]: because you cannot wait without fiber :dancer:

File: decorators.ts [hunk 9]
```diff
@@ -4,36 +4,53 @@
...
+			let result: any;
+			try {
+				result = originalMethod.apply(originalModule, args);
+			} catch(err) {
+				let promise = new Promise(function(onFulfilled : Function, onRejected: Function) {
+					onRejected(err);
 				});
-		};
 
+				// TODO: Check if we should return [promise] or promise.
```
Comment (teobugslayer) [inline decorators.ts]: This todo looks really scary - can we implement it in this PR?

Comment (rosen-vladimirov) [inline decorators.ts]: We could... In case we know what to do - I'm not sure what should be returned in this case as we do not know to return type of the function.

File: mobile/mobile-core/deviceEmitter.ts [hunk 10]
```diff
@@ -0,0 +1,35 @@
+///<reference path="./../../.d.ts"/>
+"use strict";
+
+import { EventEmitter } from "events";
+
+import fiberBootstrap = require("../../fiber-bootstrap");
```
Comment (teobugslayer) [inline mobile/mobile-core/deviceEmitter.ts]: This can be ES6-style imported as well. Not a merge stopper.

File: mobile/mobile-core/deviceEmitter.ts [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+
+			this.$androidDeviceDiscovery.on("deviceLost", (data: Mobile.IDevice) => {
+				this.emit("deviceLost", data.deviceInfo);
+			});
+
+			setInterval(() =>
+				fiberBootstrap.run(() => {
+					this.$androidDeviceDiscovery.startLookingForDevices().wait();
+				}),
+			500).unref();
```
Comment (teobugslayer) [inline mobile/mobile-core/deviceEmitter.ts]: This 500 should be a named constant. I am guilty of introducing this magic number :(

**Comment by justcodebuilduser** [comment]
:heart:

**Comment by justcodebuilduser** [comment]
:heart:

File: mobile/mobile-core/android-device-discovery.ts (hunk 2 above)
Comment (teobugslayer) [inline mobile/mobile-core/android-device-discovery.ts]: Maybe we can pre-compute this path (in the constructor) and change this property to return the cached value? We can rename it and the usage would be `${this.$staticConfig.adbPath}`

File: commands/device/run-application.ts [hunk 12]
```diff
@@ -13,17 +13,14 @@ export class RunApplicationOnDeviceCommand implements ICommand {
 
 	public execute(args: string[]): IFuture<void> {
 		return (() => {
-			this.$devicesServices.initialize({ deviceId: this.$options.device, skipInferPlatform: true }).wait();
+			this.$devicesService.initialize({ deviceId: this.$options.device, skipInferPlatform: true }).wait();
 
-			if (this.$devicesServices.deviceCount > 1) {
-				this.$errors.fail("More than one device found. Specify device explicitly with --device option. " +
-					"To discover device ID, use $%s device command.", this.$staticConfig.CLIENT_NAME.toLowerCase());
+			if (this.$devicesService.deviceCount > 1) {
```
Comment (teobugslayer) [inline commands/device/run-application.ts]: There's a missing space after the full stop. Is it possible this message to be shown in the context of Proton? It would be very weird there.

File: decorators.ts [hunk 13]
```diff
@@ -3,35 +3,69 @@
...
+			let originalMethod: Function = originalModule[propertyKey];
+			let result: any;
+			try {
+				result = originalMethod.apply(originalModule, args);
+			} catch(err) {
+				let promise = new Promise(function(onFulfilled : Function, onRejected: Function) {
+					onRejected(err);
 				});
+
+				// TODO: Check if we should return [promise] or promise.
```
Comment (teobugslayer) [inline decorators.ts]: We talked in person with @rosen-vladimirov and decided that the better approach is to throw the err instead. This way, we keep the intent of the original code and allow the high-level client code to take the correct action.

File: decorators.ts [hunk 14]
```diff
@@ -3,35 +3,69 @@
...
+				return promise;
+			}
+
+			let types = _(result)
+						.groupBy(f => typeof f)
+						.keys()
+						.value();
+
+			// Check if method returns IFuture<T>[]. In this case we will return Promise<T>[]
+			if(_.isArray(result) && types.length === 1 && typeof(_.first<any>(result).wait) === "function") {
```
Comment (teobugslayer) [inline decorators.ts]: The check whether a value is a future happens on multiple places. Is it profitable to extract it in a dedicated function?

File: fiber-bootstrap.ts [hunk 15]
```diff
@@ -3,14 +3,20 @@
...
+	if(Fiber.current) {
+		// Use the already existing fiber, we do not need new one.
 		action();
-		$injector.dispose();
-		Future.assertNoFutureLeftBehind();
-	}).run();
+	} else {
+		Fiber(() => {
+			// TODO: Think about correct place for this.
+			// errors.installUncaughtExceptionListener();
```
Comment (teobugslayer) [inline fiber-bootstrap.ts]: I think we can change errors.installUncaughtExceptionListener to be idempotent, this make it safe to be called many times. Internally, we can use _.once with the code of the existing function. After that,w e can uncomment the currently commented out code.

File: mobile/mobile-core/android-device-discovery.ts [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+			let result = this.$childProcess.exec(requestAllDevicesCommand).wait();
+
+			let currentDevices = result.toString().split(EOL).slice(1)
+				.filter( (element:string) => !helpers.isNullOrWhitespace(element) )
+				.map((element: string) => {
+					// http://developer.android.com/tools/help/adb.html#devicestatus
+					let parts = element.split("\t");
+					let identifier = parts[0];
+					let state = parts[1];
```
Comment (teobugslayer) [inline mobile/mobile-core/android-device-discovery.ts]: Can you use destructuring here? `let [identifier, state] = elements.split()`

File: mobile/mobile-core/android-device-discovery.ts [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+				newDevices = _.difference(currentDevices, this._devices);
+
+			_.each(newDevices, d => this.createAndAddDevice(d));
+			_.each(oldDevices, d => this.deleteAndRemoveDevice(d));
+		}).future<void>()();
+	}
+
+	public ensureAdbServerStarted(): IFuture<void> {
+		let startAdbServerCommand = `"${this.$staticConfig.getAdbFilePath().wait()}" start-server`;
+		return this.$childProcess.exec(startAdbServerCommand);
```
Comment (teobugslayer) [inline mobile/mobile-core/android-device-discovery.ts]: Can we use spawn and avoid creating an extra process?

File: mobile/mobile-core/devices-service.ts [hunk 18]
```diff
@@ -1,14 +1,15 @@
 ///<reference path="./../../.d.ts"/>
 "use strict";
 
-import * as util from "util";
+import util = require("util");
```
Comment (teobugslayer) [inline mobile/mobile-core/devices-service.ts]: Please revert this change

File: mobile/mobile-core/devices-service.ts [hunk 19]
```diff
@@ -1,14 +1,15 @@
...
 import Future = require("fibers/future");
-import * as helpers from "../../helpers";
+import helpers = require("../../helpers");
```
Comment (teobugslayer) [inline mobile/mobile-core/devices-service.ts]: Please revert this change

File: mobile/mobile-core/devices-service.ts [hunk 20]
```diff
@@ -1,14 +1,15 @@
...
 let assert = require("assert");
-import * as constants from "../constants";
+import constants = require("../constants");
```
Comment (teobugslayer) [inline mobile/mobile-core/devices-service.ts]: Please revert this change

File: mobile/mobile-core/devices-service.ts (hunk 19 above)
Comment (rosen-vladimirov) [inline mobile/mobile-core/devices-service.ts]: poor rebasing skills :crying_cat_face:

File: commands/device/run-application.ts [hunk 21]
```diff
@@ -13,17 +13,14 @@ export class RunApplicationOnDeviceCommand implements ICommand {
...
+				this.$errors.fail("More than one device found. Specify device explicitly with --device option.To discover device ID, use $%s device command.", this.$staticConfig.CLIENT_NAME.toLowerCase());
```
Comment (Fatme) [inline commands/device/run-application.ts]: `failWithoutHelp` maybe?

File: commands/device/run-application.ts [hunk 22]
```diff
@@ -13,17 +13,14 @@ export class RunApplicationOnDeviceCommand implements ICommand {
...
 			}
 
-			let action = (device: Mobile.IDevice) => {
-				return (() => device.applicationManager.startApplication(args[0]).wait()).future<void>()();
-			};
-			this.$devicesServices.execute(action).wait();
+			let action = (device: Mobile.IDevice) =>  { return (() => device.applicationManager.startApplication(args[0]).wait()).future<void>()(); };
```
Comment (Fatme) [inline commands/device/run-application.ts]: `let action = (device: Mobile.IDevice) => device.applicationManager.startApplication(args[0])` - No need to create an extra future.

File: commands/device/uninstall-application.ts [hunk 23]
```diff
@@ -10,12 +10,10 @@ export class UninstallApplicationCommand implements ICommand {
...
 	public execute(args: string[]): IFuture<void> {
 		return (() => {
-			this.$devicesServices.initialize({ deviceId: this.$options.device, skipInferPlatform: true }).wait();
+			this.$devicesService.initialize({ deviceId: this.$options.device, skipInferPlatform: true }).wait();
 
-			let action = (device: Mobile.IDevice) =>  {
-				return (() => device.applicationManager.uninstallApplication(args[0]).wait()).future<void>()();
-			};
-			this.$devicesServices.execute(action).wait();
+			let action = (device: Mobile.IDevice) =>  { return (() => device.applicationManager.uninstallApplication(args[0]).wait()).future<void>()(); };
```
Comment (Fatme) [inline commands/device/uninstall-application.ts]: Use `let action = (device: Mobile.IDevice) => device.applicationManager.uninstallApplication(args[0]);` - No need to create an extra future.

**Comment by justcodebuilduser** [comment]
:heart:

**Comment by justcodebuilduser** [comment]
:heart:

**Comment by justcodebuilduser** [comment]
:heart:

File: bootstrap.ts [hunk 24]
```diff
@@ -95,3 +94,6 @@ $injector.requireCommand("doctor", "./commands/doctor");
 
 $injector.require("utils", "./utils");
 $injector.require("bplistParser", "./bplist-parser");
+
+import errors = require("./errors");
+errors.installUncaughtExceptionListener();
```
Comment (rosen-vladimirov) [inline bootstrap.ts]: remove this change

**Comment by justcodebuilduser** [comment]
:heart:

File: bootstrap.ts [hunk 25]
```diff
@@ -95,3 +94,13 @@ $injector.requireCommand("doctor", "./commands/doctor");
...
 $injector.require("utils", "./utils");
 $injector.require("bplistParser", "./bplist-parser");
+
+$injector.requirePublicClass("deviceEmitter", "./mobile/mobile-core/deviceEmitter");
+
+$injector.register("mobilePlatformsCapabilities", {});
+$injector.register("config", {});
+$injector.register("analyiticsService", {});
+$injector.require("staticConfig", "./static-config-base");
+$injector.requirePublicClass("deviceLogProvider", "./mobile/mobile-core/device-log-provider");
```
Comment (Mitko-Kerezov) [inline bootstrap.ts]: this class shouldn't necessarily be publicly visible

File: mobile/ios/ios-core.ts [hunk 26]
```diff
@@ -669,15 +670,26 @@ class WinSocket implements Mobile.IiOSDeviceSocket {
...
-			printData(data);
+			let output = ref.readCString(data, 0);
+			process.send(output);
 			data = this.read(WinSocket.BYTES_TO_READ);
 		}
 		this.close();
 	}
 
+	public readSystemLog(printData: any) {
+		let ch = require("child_process");
```
Comment (Mitko-Kerezov) [inline mobile/ios/ios-core.ts]: We can think about a mockable dependency here instead of requiring node's child_process module directly

**Comment by justcodebuilduser** [comment]
:broken_heart:

**Comment by justcodebuilduser** [comment]
:broken_heart:

**Comment by justcodebuilduser** [comment]
:heart:

File: commands/device/run-application.ts (hunk 12 above)
Comment (rosen-vladimirov) [inline commands/device/run-application.ts]: fixed

File: bootstrap.ts [hunk 27]
```diff
@@ -96,3 +95,13 @@ $injector.requireCommand("doctor", "./commands/doctor");
 
 $injector.require("utils", "./utils");
 $injector.require("bplistParser", "./bplist-parser");
+
+$injector.requirePublicClass("deviceEmitter", "./mobile/mobile-core/device-emitter");
+
+$injector.register("mobilePlatformsCapabilities", {});
+$injector.register("config", {});
```
Comment (rosen-vladimirov) [inline bootstrap.ts]: maybe it's better to register mock classes and require the files here instead of registering them

File: commands/device/run-application.ts [hunk 28]
```diff
@@ -13,17 +13,14 @@ export class RunApplicationOnDeviceCommand implements ICommand {
...
-					"To discover device ID, use $%s device command.", this.$staticConfig.CLIENT_NAME.toLowerCase());
+			if (this.$devicesService.deviceCount > 1) {
+				this.$errors.failWithoutHelp("More than one device found. Specify device explicitly with --device option. To discover device ID, use $%s device command.", this.$staticConfig.CLIENT_NAME.toLowerCase());
 			}
 
-			let action = (device: Mobile.IDevice) => {
-				return (() => device.applicationManager.startApplication(args[0]).wait()).future<void>()();
-			};
-			this.$devicesServices.execute(action).wait();
+			let action = (device: Mobile.IDevice) => device.applicationManager.startApplication(args[0])
```
Comment (rosen-vladimirov) [inline commands/device/run-application.ts]: missing `;` at the end

File: errors.ts [hunk 29]
```diff
@@ -64,26 +64,38 @@ function resolveCallStack(error: Error): string {
...
+						callstack = resolveCallStack(err);
+					} catch(err) {
+						console.error("Error while resolving callStack:", err);
+					}
+				}
+				console.error(callstack || err.toString());
+
+				if(!$injector.resolve("staticConfig").disableAnalytics) {
+					try {
+						let analyticsService = $injector.resolve("analyticsService");
```
Comment (rosen-vladimirov) [inline errors.ts]: this code will fail on proton side as analyticsService is mocked, we should consider adding real code here

File: mobile/ios/ios-core.ts [hunk 30]
```diff
@@ -669,15 +671,25 @@ class WinSocket implements Mobile.IiOSDeviceSocket {
...
+			process.send(output);
 			data = this.read(WinSocket.BYTES_TO_READ);
 		}
 		this.close();
 	}
 
+	public readSystemLog(printData: any): void {
+		let serviceArg: number|string = this.service || '';
+		let formatArg: number|string = this.format || '';
+		let sysLog = this.$childProcess.fork(path.join(__dirname, "ios-sys-log.js"), [serviceArg.toString(), formatArg.toString()], {silent: true});
```
Comment (rosen-vladimirov) [inline mobile/ios/ios-core.ts]: is the `.js` mandatory for `ios-sys-log`

File: mobile/mobile-core/device-emitter.ts [hunk 31]
```diff
@@ -0,0 +1,51 @@
...
+"use strict";
+
+import { EventEmitter } from "events";
+import byline = require("byline");
+import fiberBootstrap = require("../../fiber-bootstrap");
+
+class DeviceEmitter extends EventEmitter {
+	private static DEVICE_DISCOVERY_TIMEOUT = 500;
+
+	// TODO: add iOSDeviceDiscovery as a dependency too
```
Comment (rosen-vladimirov) [inline mobile/mobile-core/device-emitter.ts]: remove this TODO

**Comment by justcodebuilduser** [comment]
:heart:

**Comment by justcodebuilduser** [comment]
:heart:

**Comment by justcodebuilduser** [comment]
:heart:

**Comment by justcodebuilduser** [comment]
:heart:

File: README.md [hunk 32]
```diff
@@ -116,7 +117,7 @@ Usage
 In order to use mobile-cli-lib, just add a reference to it in your package.json:
 ```JSON
 dependencies: {
-	"mobile-cli-lib": "https://github.com/telerik/mobile-cli-lib/tarball/master"
+	"mobile-cli-lib": "0.0.2"
```
Comment (teobugslayer) [inline README.md]: rather it is `0.0.4`

File: appbuilder/proton-bootstrap.ts [hunk 33]
```diff
@@ -0,0 +1,15 @@
+///<reference path="../.d.ts"/>
+"use strict";
+
+require("../bootstrap");
+
+import {OptionsBase} from "../options";
+
+$injector.require("staticConfig", "./appbuilder/proton-static-config");
+// TODO: Add real dependencies
```
Comment (teobugslayer) [inline appbuilder/proton-bootstrap.ts]: Well, lets's add some real dependencies! or remove the comment.

File: child-process.ts [hunk 34]
```diff
@@ -49,6 +49,11 @@ export class ChildProcess implements IChildProcess {
 		return child_process.spawn(command, args, options);
 	}
 
+	public fork(modulePath: string, args?: string[], options?: any): child_process.ChildProcess {
+		this.$logger.debug("fork: %s %s", modulePath, args.join(" "));
```
Comment (teobugslayer) [inline child-process.ts]: This log does not add double quotes around arguments with spaces, which makes debugging very hard. We already have this issue in other child process functions and had to use other methods to get the correct command line.

File: errors.ts (hunk 29 above)
Comment (teobugslayer) [inline errors.ts]: :+1:

File: errors.ts [hunk 35]
```diff
@@ -64,26 +64,38 @@ function resolveCallStack(error: Error): string {
...
-			try {
-				let analyticsService = $injector.resolve("analyticsService");
-				analyticsService.trackException(err, callstack);
-			} catch (e) {
-				// Do not replace with logger due to cyclic dependency
-				console.error("Error while reporting exception: " + e);
-			}
-		}
+				// TODO: Find what to do here for CLIs
+				// process.exit(ErrorCodes.UNKNOWN);
```
Comment (teobugslayer) [inline errors.ts]: a very scary comment

File: mobile/mobile-core/android-device-discovery.ts [hunk 36]
```diff
@@ -0,0 +1,100 @@
+///<reference path="./../../.d.ts"/>
+"use strict";
+
+import {DeviceDiscovery} from "./device-discovery";
+import * as helpers from "../../helpers";
+import {AndroidDevice} from "../android/android-device";
+import {EOL} from "os";
+import Future = require("fibers/future");
+import fiberBootstrap = require("../../fiber-bootstrap");
```
Comment (teobugslayer) [inline mobile/mobile-core/android-device-discovery.ts]: can we use ES6 import here?

File: errors.ts (hunk 35 above)
Comment (rosen-vladimirov) [inline errors.ts]: Pfff, I already know what to do, but I forgot to add it in this PR.
Thanks for pointing it :) I'll fix it

File: package.json [hunk 37]
```diff
@@ -32,13 +32,13 @@
     "cli-table": "https://github.com/telerik/cli-table/tarball/v0.3.1.1",
     "colors": "0.6.2",
     "cookie": "0.1.0",
-    "ffi": "https://github.com/icenium/node-ffi/tarball/v1.3.1.0",
-    "fibers": "https://github.com/icenium/node-fibers/tarball/v1.0.6.0",
+    "ffi": "https://github.com/icenium/node-ffi/tarball/v1.3.2.0-proton",
+    "fibers": "https://github.com/icenium/node-fibers/tarball/v1.0.6.0-proton",
```
Comment (teobugslayer) [inline package.json]: Should we use 1.0.6.2 instead?

File: test/unit-tests/decorators.ts [hunk 38]
```diff
@@ -133,5 +133,50 @@ describe("decorators", () => {
...
+			// Call this line in order to generate publicApi and get the real Promises
+			promisifiedResultFunction({}, "propertyName", {});
+			let promises: Promise<string>[] = $injector.publicApi.__modules__["moduleName"]["propertyName"]();
+			_.each(promises, (promise, index) => promise.then((val: string) => {
+				assert.deepEqual(val, expectedResults[index]);
+			}, (err: Error) => {
+				// We cannot compare promise.reason() with error directly as node-fibers modify the error.stack property, so deepEqual method fails.
+				assert.deepEqual(err.message, expectedResults[index].message);
+			}));
+		});
```
Comment (teobugslayer) [inline test/unit-tests/decorators.ts]: :100: for the tests!

File: package.json (hunk 37 above)
Comment (rosen-vladimirov) [inline package.json]: no, Proton is using different branch of fibers

File: errors.ts (hunk 29 above)
Comment (rosen-vladimirov) [inline errors.ts]: after discussion, we've decided to set `disableAnalytics` to true for Proton, so it common-lib will not track the exceptions. Proton will do it on its own.

**Comment by justcodebuilduser** [comment]
:broken_heart:

**Comment by justcodebuilduser** [comment]
:heart:

File: package.json (hunk 37 above)
Comment (teobugslayer) [inline package.json]: ok

**Comment by teobugslayer** [comment]
:+1:
